### PR TITLE
[server][da-vinci] Fix regression in recent ingestion isolation improvements.

### PR DESCRIPTION
**Pull request by sixpluszero**
## [server][da-vinci] Fix regression in recent ingestion isolation improvements

This PR fixes recent regression added to the ingestion isolation command request logics.
1. Do not throw exception when remote ingestion process reject command as it should be a transient state during unsubscription from II to main process. (improvement)
2. Try to initialize topic partition status to be subscribed in the forked process before START_CONSUMPTION command execution check. If the topic partition has not been touched in the forked process this action will initialize the state so ingestion command can go through (bug fix)
3. For any commands other than START_CONSUMPTION, if a topic partition is not hosted in a server, the default go-to process should be main process instead of forked process. (bug fix)

Local test fixes the TestBatchForIngestionIsolation test failures and hanging in the cleanup() method. Running CI to verify the fix.

## How was this PR tested?

Submitted to multiple CI tests.
## Does this PR introduce any user-facing changes?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/isolated/IsolatedIngestionServer.java
Comment (FelixGV) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/isolated/IsolatedIngestionServer.java:446]: Nitpick: the code in this function is pretty verbose... can we consider doing it more succinctly, e.g.:

```
    topicPartitionSubscriptionMap.computeIfAbsent(topicName, s -> new VeniceConcurrentHashMap<>())
        .computeIfAbsent(partition, p -> new AtomicBoolean(true))
        .set(true);
```

Same comment for `setResourceToBeUnsubscribed`

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java
Comment (FelixGV) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java:273]: Nitpick: this kind of multi-part boolean equation can be a bit hard to read... it may help readability to reduce it a bit. Some suggestions include:

1. The third boolean is wrapped in unnecessary parenthesis, which can be removed.
2. Adding a static import for `IngestionCommandType.START_CONSUMPTION` would allow you to write just `START_CONSUMPTION` which is sufficiently precise to understand, but shorter.
3. Enums don't need `equals` and can be compared with `==` (or in this case, `!=`, since you're negating).

End result could be something like:

```
      if (isTopicPartitionHostedInMainProcess(topicName, partition) || (!isTopicPartitionIngesting(topicName, partition)
           && command != START_CONSUMPTION)) {
```
Or (if spotless allows is), perhaps:

```
      if (isTopicPartitionHostedInMainProcess(topicName, partition) 
            || (!isTopicPartitionIngesting(topicName, partition) && command != START_CONSUMPTION)) {
```

BTW, can you think of a way for this logic change to be unit testable?

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java
Comment (FelixGV) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java:155]: Why does the ordering of these statements matter? I don't have any concern, but am just curious.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/isolated/IsolatedIngestionServer.java
Comment (FelixGV) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/isolated/IsolatedIngestionServer.java:470]: Since we are not going to use `subscription in any way, this would be a good use case for `containsKey`...

```
return getTopicPartitionSubscriptionMap().getOrDefault(topicName, Collections.emptyMap()).containsKey(partition);
```

Possibly put `containsKey` on the next line, if you'd like.

**Review by FelixGV — COMMENTED** [review]
Thanks a lot Jialin. LGTM. Only nitpicks...

File: internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestBatch.java
Comment (sixpluszero) [reply internal/venice-test-common/src/integrationtest/java/com/linkedin/venice/endToEnd/TestBatch.java:474]: Initially I saw this test failed during a hang run, but it turns out that it might just be flaky. Reverted. TestBatch has 1-2 tests flaky, but we can check back later as the top priority is to fix the build.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java
Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java:155]: With the final fix,  it does not matter. This change happens during initial debugging, but I think either place should be fine.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/isolated/IsolatedIngestionServer.java
Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/isolated/IsolatedIngestionServer.java:470]: Actually this method is not used after final fix. Just removed it.

Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/isolated/IsolatedIngestionServer.java:470]: Even better :D

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java
Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java:157]: If it does not matter then I would lean towards not changing it. This way the git history is easier to understand. But it's up to you.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java
Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java]: It could be. Let me issue a bunch of wc-tests after address those comments to make sure no more regression and I will add unit test at the same time.

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java
Comment (FelixGV) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java:3]: In general we want to avoid wildcard imports, although for the static import of enum values I guess the risk of misuse should be pretty low... Eventually we may implement a static analysis rule to prevent all wildcard imports though, in which case that would likely end up getting changed.

Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackend.java:3]: I did not mean to do that but I found that IntelliJ does not save the trick config we use to avoid static import every time it shutdowns.. will change back.

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackendTest.java
Comment (FelixGV) [inline clients/da-vinci-client/src/test/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackendTest.java:24]: Did you copy this test on the main branch and confirm that it failed?

**Review by FelixGV — APPROVED** [review]
Thanks Jialin! LGTM! Assuming the CI still looks good after these changes, feel free to merge!

File: clients/da-vinci-client/src/test/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackendTest.java
Comment (sixpluszero) [reply clients/da-vinci-client/src/test/java/com/linkedin/davinci/ingestion/IsolatedIngestionBackendTest.java:24]: yes it is.

**Pull request merged**
